addons/pos_online_payment/controllers/payment_portal.py

```diff
@@ -0,0 +1,302 @@
+
+        user_sudo = request.env.user
+        logged_in = not user_sudo._is_public()
+        partner_sudo = self._get_partner_sudo(user_sudo)
+        if not partner_sudo:
+            return self._redirect_login(request)
+
+        kwargs = {
+            'pos_order_id': pos_order_sudo.id,
+        }
+        exit_route_arg = self._get_exit_route_arg(exit_route)
+        rendering_context = {
+            **kwargs,
+            'exit_route': exit_route,
+            'reference_prefix': request.env['payment.transaction'].sudo()._compute_reference_prefix(provider_code=None, separator='-', **kwargs),
+            'partner_id': partner_sudo.id,
+            'access_token': access_token,
+            'transaction_route': f'/pos/pay/transaction/{pos_order_sudo.id}?access_token={access_token}{exit_route_arg}',
+            'landing_route': self._get_landing_route(pos_order_sudo.id, access_token, exit_route_arg=exit_route_arg),
+            **self._get_custom_rendering_context_values(**kwargs),
+        }
+
+        currency_id = pos_order_sudo.currency_id
+
+        if not self._is_valid_currency(currency_id):
+            rendering_context['currency'] = False
+            return self._render_pay(request, rendering_context)
+        rendering_context['currency'] = currency_id
+
+        amount_to_pay = self._get_amount_to_pay(pos_order_sudo)
+        if not self._is_valid_amount(amount_to_pay, currency_id):
+            rendering_context['amount'] = False
+            return self._render_pay(request, rendering_context)
+        rendering_context['amount'] = amount_to_pay
+
+        providers_sudo = self._get_allowed_providers_sudo(
+            request, pos_order_sudo, partner_sudo.id, amount_to_pay)
+
+        # Compute the fees taken by providers supporting the feature
+        fees_by_provider = {
+            provider_sudo: provider_sudo._compute_fees(
+                amount_to_pay, currency_id, partner_sudo.country_id)
+            for provider_sudo in providers_sudo.filtered('fees_active')
+        }
+
+        if logged_in:
+            tokens_sudo = request.env['payment.token']._get_available_tokens(
+                providers_sudo.ids, partner_sudo.id)
+            show_tokenize_input = self._compute_show_tokenize_input_mapping(
+                providers_sudo, **kwargs)
+        else:
+            tokens_sudo = False
+            show_tokenize_input = {p.id: False for p in providers_sudo}
+
+        rendering_context.update({
+            'providers': providers_sudo,
+            'tokens': tokens_sudo,
+            'fees_by_provider': fees_by_provider,
+            'show_tokenize_input': show_tokenize_input,
+            **self._get_custom_rendering_context_values(**kwargs),
+        })
+        return self._render_pay(request, rendering_context)
+
+    @staticmethod
+    def _render_pay(request, rendering_context):
+        return request.render('pos_online_payment.pay', rendering_context)
```

Comment:
In this file, is there a way to generate custom `request` object? If all the `request` object here is the same, we can consider it global. No need to pass it around method/functions.

Reply:
I don't really know how that "request" object works, so I don't know if it is safe to use it in a staticmethod. For me, it makes sense that the "request" variable is linked to the current request of an "@http.route" method, but it doesn't make sense that the "request" variable would be linked to the correct request when the "@staticmethod" would be called. But it's only my assumption, I don't know well Python nor the Odoo API.

Reply:
Ok so I assume it is safe to call it from an @http.route method and use request in it without passing it as a parameter.
This method avoids copy/paste. If tomorrow the name of the "pay" template changes, or the way the context is passed to it is changed, only this method would need to be changed.